File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> leftMaxHeight, rightMaxHeight;
        int h = 0, n = height.size();
        for(int i=0;i<n;i++) {
            h = max(h, height[i]);
            leftMaxHeight.push_back(h);
        }
        
        h = 0;
        for(int i=n-1;i>=0;i--) {
            h = max(h, height[i]);
            rightMaxHeight.push_back(h);
        }
        reverse(rightMaxHeight.begin(), rightMaxHeight.end());
        
        int ans = 0;
        cout << ans << " ";
        for(int i=1;i<n-1;i++) {
            int water = min(leftMaxHeight[i], rightMaxHeight[i]) - height[i];
            if(water > 0) ans += water;
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = 0, right = height.length-1;
    let leftMax = 0, rightMax = 0, ans = 0;
    
    while(left <= right) {
        if(height[left] <= height[right]) {
            if(height[left] >= leftMax) leftMax = height[left];
            else ans += leftMax - height[left];
            left += 1;
        } else {
            if(height[right] >= rightMax) rightMax = height[right];
            else ans += rightMax - height[right];
            right -= 1;
        }
    }
    return ans;
};